File: minimum-genetic-mutation/minimum-genetic-mutation.java
class Solution {
    public int minMutation(String startGene, String endGene, String[] bank) {
        Queue<String> queue = new ArrayDeque<>();
        Set<String> seen = new HashSet<>();
        queue.add(startGene);
        seen.add(startGene);
        int steps = 0;

        while (!queue.isEmpty()) {
            int size = queue.size();
            for (int j = 0; j < size; j++) {
                String node = queue.poll();
                if (node.equals(endGene)) {
                    return steps;
                }

                for (char c : new char[] {'A', 'C', 'G', 'T'}) {
                    for (int i = 0; i < node.length(); i++) {
                        String neighbor = node.substring(0, i) + c + node.substring(i + 1);
                        if (!seen.contains(neighbor) && Arrays.asList(bank).contains(neighbor)) {
                            queue.add(neighbor);
                            seen.add(neighbor);
                        }
                    }
                }
            }
            steps++;
        }

        return -1;
    }
}

File: 433-minimum-genetic-mutation/minimum-genetic-mutation.py
class Solution:
    def minMutation(self, startGene: str, endGene: str, bank: List[str]) -> int:
        geneBank = set(bank)
        visited = {startGene}
        q = deque([startGene])
        steps = 0

        while q:
            size = len(q)
            for i in range(size):
                curr = q.popleft()
                if curr == endGene:
                    return steps
                for i in range(8):
                    for char in ('A', 'C', 'G', 'T'):
                        newString = curr[:i] + char + curr[i + 1:]
                        if newString in geneBank and newString not in visited: 
                            q.append(newString)
                            visited.add(newString)
            steps += 1
        return -1